knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java

```diff
@@ -51,19 +56,41 @@
         idx = matchResult.end();
       }
       if (idx < html.length()) {
-        fragments.add(toRaw(html, idx, html.length()));
+        processMarkup(fragments, html, idx, html.length());
       }
     } else {
-      fragments.add(toRaw(html, 0, html.length()));
+      processMarkup(fragments, html, 0, html.length());
     }
     return fragments;
   }
 
-  private Fragment toRaw(String html, int startIdx, int endIdx) {
-    return Fragment.raw(html.substring(startIdx, endIdx));
+  private void processMarkup(List<Fragment> fragments, String data, int startIdx, int endIdx) {
```

Comment:
Now, for each found fragment, searching for fallbacks is triggered and searching for matching fallback is done from the beginning. Wouldn't be more efficient, to build some cache/map of fallbacks first and then only add proper fallback (if exists) by name?

Reply:
Introduced FallbackMarkers to run the fallback matcher once against the whole html